File: plugins/DEV-ESSENTIALS/hooks/block-no-verify.js
#!/usr/bin/env node
/**
 * PreToolUse Hook: Block --no-verify
 *
 * Prevents `git commit --no-verify` which bypasses pre-commit hooks.
 *
 * Exit codes:
 *   0 - Allow (not a --no-verify commit)
 *   2 - Block (--no-verify detected)
 */

const MAX_STDIN = 1024 * 1024;

let data = '';
process.stdin.setEncoding('utf8');

process.stdin.on('data', chunk => {
  if (data.length < MAX_STDIN) data += chunk.substring(0, MAX_STDIN - data.length);
});

process.stdin.on('end', () => {
  try {
    const disabled = (process.env.DISABLED_HOOKS || '').split(',');
    if (disabled.includes('block-no-verify')) {
      process.stdout.write(data);
      return;
    }

    const input = JSON.parse(data);
    const command = input.tool_input?.command || '';

    if (command.includes('git commit') && command.includes('--no-verify')) {
      console.error('[Hook] BLOCKED: git commit --no-verify is not allowed.');
      console.error('[Hook] Pre-commit hooks exist to protect code quality. Remove --no-verify and fix any issues.');
      process.stdout.write(data);
      process.exit(2);
    }

    process.stdout.write(data);
  } catch (err) {
    console.error(`[Hook] block-no-verify error: ${err.message}`);
    process.stdout.write(data);
  }
});

File: plugins/DEV-ESSENTIALS/hooks/block-no-verify.py
#!/usr/bin/env python3
"""
PreToolUse Hook: Block --no-verify

Prevents `git commit --no-verify` which bypasses pre-commit hooks.

Exit codes:
  0 - Allow (not a --no-verify commit)
  2 - Block (--no-verify detected)
"""

import json
import os
import sys

MAX_STDIN = 1024 * 1024

def main():
    data = sys.stdin.read(MAX_STDIN)

    try:
        disabled = os.environ.get("DISABLED_HOOKS", "").split(",")
        if "block-no-verify" in disabled:
            sys.stdout.write(data)
            return

        inp = json.loads(data)
        command = inp.get("tool_input", {}).get("command", "")

        if "git commit" in command and "--no-verify" in command:
            print("[Hook] BLOCKED: git commit --no-verify is not allowed.", file=sys.stderr)
            print("[Hook] Pre-commit hooks exist to protect code quality. Remove --no-verify and fix any issues.", file=sys.stderr)
            sys.stdout.write(data)
            sys.exit(2)

        sys.stdout.write(data)
    except Exception as err:
        print(f"[Hook] block-no-verify error: {err}", file=sys.stderr)
        sys.stdout.write(data)

if __name__ == "__main__":
    main()
